<?xml version="1.0" encoding="UTF-8" ?>
<launch>
    <!-- args -->
    <arg name="use_rviz" default="true"/>            #### 动态传参，使用rviz可视化工具
    <arg name="scan_topic" default="scan"/>          #### 原始laser话题
    <arg name="filtered_topic" default="scan_filtered"/> #### 滤波后laser话题
    <arg name="scan_2_pc_topic" default="scan_2_pc"/>#### 转换为点云后的话题名
    <arg name="map_frame" default="map"/>            #### 地图坐标系frame
    <arg name="odom_frame" default="odom"/>          #### 里程计坐标系frame
    <arg name="base_frame" default="base_link"/>     #### 移动平台坐标系frame
    <arg name="publish_covariance" default="false"/> #### 动态传参
    <param name="use_sim_time" value="true"/>        #### 公共空间的参数，使用仿真时间

    <!-- static_transform_publisher， 静态坐标系转换发布 -->
    #### publish an example base_link -> laser_frame transform #####
    <node pkg="tf2_ros" type="static_transform_publisher" name="base_link2laser_frame"
          args="0 0 0 0 0 0 base_link laser_frame" output="screen"/>

    <!-- play，播放bag包 -->
    <node pkg="rosbag" type="play" name="play" required="false"
          args="$(find multi_level_lidar_map)/hokuyo_laser_scan_0531.bag --delay=10 --clock">
    </node>

    <!-- scan_to_scan_filter_chain，启动滤波器 -->
    <node pkg="laser_filters" type="scan_to_scan_filter_chain" name="laser_filter">
        #### 加载滤波器参数 ######
        <rosparam command="load" file="$(find multi_level_lidar_map)/config/hokuyo_filter.yaml" />
        <!-- remap from="scan" to="base_scan" -->
    </node>

    <!-- laser_scan_matcher_node，启动配准,提供里程计 -->
    #### start the laser scan_matcher ##############################
    <group if="$(arg publish_covariance)">
        <param name="laser_scan_matcher_node/do_compute_covariance" value="1"/>
        <param name="laser_scan_matcher_node/publish_pose_with_covariance" value="true"/>
        <param name="laser_scan_matcher_node/publish_pose_with_covariance_stamped" value="true"/>
    </group>

    <node pkg="laser_scan_matcher" type="laser_scan_matcher_node" name="laser_scan_matcher_node" output="screen">
        #### 固定坐标系 #########
        <param name="fixed_frame" value="$(arg odom_frame)"/> #### 固定坐标系，节点私有参数
        <param name="base_frame" value="$(arg base_frame)"/>  #### 激光雷达坐标系
        #### 可使用的额外信息 ####
        <param name="use_imu" value="false"/>         #### 是否有可用的imu
        <param name="use_odom" value="false"/>        #### 是否有可用的odom
        <param name="use_vel" value="false"/>         #### 是否有可用的vel
        #### 发布项目 ###########
        <param name="publish_pose" value="false"/>        #### 发布位姿
        <param name="publish_pose_stamped" value="true"/> #### 发布带有时间戳的位姿
        #### 关键帧要求 #########
        <param name="kf_dist_linear" value="0.1"/>    #### 关键帧需要移动的距离
        <param name="kf_dist_angular" value="0.175"/> #### 关键帧需要变动的角度,0.175rad,10度
        #### icp参数配置 ########
        <param name="max_iterations" value="20"/>     #### 最大迭代次数
    </node>

    <!-- scan_to_cloud_converter_node，laser转point cloud，提供三维点云数据 -->
    <node pkg="scan_to_cloud_converter" type="scan_to_cloud_converter_node" name="scan_to_cloud_converter_node" output="screen">
        <param name="sub_scan_topic" value="$(arg filtered_topic)"/>  #### 接受的laser话题
        <param name="pub_pc_topic" value="$(arg scan_2_pc_topic)"/>   #### 发布的点云话题
    </node>

    <!-- hector_mapping，建图对比，用hector_mapping生成栅格地图 -->
    <node pkg="hector_mapping" type="hector_mapping" name="hector_mapping" output="screen">
        #### Frame names #########
        <param name="map_frame" value="$(arg map_frame)"/>      #### map frame
        <param name="base_frame" value="$(arg base_frame)"/>    #### base frame
        <param name="odom_frame" value="$(arg odom_frame)"/>    #### odom frame
        #### Tf use ##############
        <param name="use_tf_scan_transformation" value="true"/> #### using the tf tree to find the transform between the base frame and laser frame
        <param name="use_tf_pose_start_estimate" value="false"/>#### using the tf tree 估计初始位姿
        <param name="pub_map_odom_transform" value="true"/>     #### 确定系统是否应发布 map->odom 变换
        #### Map size / start point ##############
        <param name="map_resolution" value="0.050"/>            #### 地图精度
        <param name="map_size" value="512"/>                    #### 地图大小
        <param name="map_start_x" value="0.5"/>                 #### 启始位置x
        <param name="map_start_y" value="0.5"/>                 #### 启始位置y
        <param name="map_multi_res_levels" value="2"/>          #### 地图多分辨率网格级别的数量
        #### Map update parameters ##############
        <param name="map_pub_period" value="0.1"/>              #### 地图更新周期
        <param name="update_factor_free" value="0.4"/>          #### 用于更新范围 [0.0, 1.0] 中的空闲单元格的地图更新修饰符。值 0.5 表示没有变化
        <param name="update_factor_occupied" value="0.9"/>      #### 用于更新 [0.0, 1.0] 范围内已占用单元格的地图更新修饰符。值 0.5 表示没有变化
        <param name="map_update_distance_thresh" value="0.4"/>  #### 执行地图更新的阈值
        <param name="map_update_angle_thresh" value="0.06" />   #### 执行地图更新的阈值
        <param name="laser_z_min_value" value = "-1.0" />       #### 系统要使用的激光扫描端点相对于激光扫描仪框架的最小高度 [m]。低于此值的扫描端点将被忽略
        <param name="laser_z_max_value" value = "1.0" />
        #### Advertising config ##############
        <param name="advertise_map_service" value="true"/>      #### 地图服务
        <param name="scan_subscriber_queue_size" value="5"/>    #### 接收的队列长度
        <param name="scan_topic" value="$(arg filtered_topic)"/>#### 接收的激光话题
        #### Debug parameters ##############
        <!--
          <param name="output_timing" value="false"/>
          <param name="pub_drawings" value="true"/>
          <param name="pub_debug_output" value="true"/>
        -->
        <param name="tf_map_scanmatch_transform_frame_name" value="scanmatcher_frame"/>
    </node>

    <!-- syncPosePC，同步节点，同步滤波激光与位姿，制定节点规则 -->
    <node pkg="multi_level_lidar_map" type="syncPosePC" name="syncPosePC" output="screen">
        <param name="scan_2_pc_topic" value="scan_2_pc"/>     #### 接收的laser转点云消息
        <param name="node_liner_limit" value="0.3"/>          #### 节点距离变化限制
        <param name="node_angular_limit" value="1.57"/>       #### 节点角度变化限制
        <param name="odom_child_frame_id" value="base_link"/> #### odom子坐标系id
    </node>

    <!-- createPCMap，生成点云地图 没有用 -->
    <node pkg="multi_level_lidar_map" type="createPCMap" name="createPCMap" output="screen">
    </node>

    <!-- savePose，保存关键帧节点位姿节点 -->
    <node pkg="multi_level_lidar_map" type="savePose" name="savePose" output="screen">
        <param name="pose_save_path" value="/home/RJY/catkin_ws_multi_level_map/src/multi_level_lidar_map/key_frame_pose/key_frame_pose.yaml"/>
    </node>

    <!-- saveORBPcd，保存关键帧节点的点云与描述子 -->
    <node pkg="multi_level_lidar_map" type="saveORBPcd" name="saveORBPcd" output="screen">
        <param name="orb_save_path" value="/home/RJY/catkin_ws_multi_level_map/src/multi_level_lidar_map/key_frame_orb/key_frame_orb.yaml"/>
        <param name="frame_save_path" value="/home/RJY/catkin_ws_multi_level_map/src/multi_level_lidar_map/key_frame/"/>
    </node>

    <!-- rviz，数据可视化 -->
    <group if="$(arg use_rviz)">
        <node pkg="rviz" type="rviz" name="rviz_multi_level_lidar_map" required="true"
              args="-d $(find multi_level_lidar_map)/rvizs/view_hector.rviz"/>
    </group>

</launch>
